### PR TITLE
Homer

### PR DESCRIPTION
**Pull request by lldelisle**
FOR CONTRIBUTOR:
* [X] - I have read the [CONTRIBUTING.md](https://github.com/galaxyproject/tools-iuc/blob/master/CONTRIBUTING.md) document and this tool is appropriate for the tools-iuc repo.
* [X] - License permits unrestricted use (educational + commercial)
* [X] - This PR adds a new tool or tool collection

Hi,
Here is a summary of what I did:
- New tool: `homer_annotatePeaks` which enable to annotate a list of interval with GC content (if fasta or genome from fasta table provided), information about the closest gene (ID, name) and the distance to TSS or TTS if gtf provided. If more detailed annotations relative to genes a homer annotation file can be used. This file can be created by the following tool:
- New tool: `homer_gtf_to_annotation` which will allow to generate a homer annotation file with custom distances to TSS, TTS and store 5'UTR and 3'UTR information from gtf.
- New tool: `homer_findMotifsGenome` which perform motif finding analysis both with known motifs or de novo motifs discovery from a peak file and a fasta. The fasta file can be from history or from fasta table. The first step of this analysis is to preparse the genome at a specific size to build background sequences. To save time, these background sequences can be generated before and reused, that's why I wrote a data_manager.
- New data_manager: `homer_preparse`. It prepare the genome for motif finding. Each entry has 3 main parameters: dbkey, maks (whether the small letters were masked), size.

At the code levels, I have an interrogation:
- data_manager table: I need to know which fasta file has been used to generate the preparse. I chose to store the full path in the table. I could have made a symlink and store only the basename. I don't know what is best.

There are plenty of other tools from the homer suite to implement but a lot of them would require a new datatype as proposed here: https://github.com/bgruening/galaxytools/blob/master/tools/homer/datatypes_conf.xml

Could you let me know if you are interested in supporting this tool suite?

Thanks

### CODE REVIEW

**Comment by lldelisle** [comment]
@bgruening ready for review if you have time.

**Comment by bgruening** [comment]
>     * New tool: `homer_annotatePeaks` which enable to annotate a list of interval with GC content (if fasta or genome from fasta table provided), information about the closest gene (ID, name) and the distance to TSS or TTS if gtf provided. If more detailed annotations relative to genes a homer annotation file can be used. This file can be created by the following tool:
> 
>     * New tool: `homer_gtf_to_annotation` which will allow to generate a homer annotation file with custom distances to TSS, TTS and store 5'UTR and 3'UTR information from gtf.
> 
>     * New tool: `homer_findMotifsGenome` which perform motif finding analysis both with known motifs or de novo motifs discovery from a peak file and a fasta. The fasta file can be from history or from fasta table. The first step of this analysis is to preparse the genome at a specific size to build background sequences. To save time, these background sequences can be generated before and reused, that's why I wrote a data_manager.


That is super cool, we had so many users over the year that wanted to use Homer and we did not manage to get it in and invented workarounds ... :) Thanks a lot!
 
>     * New data_manager: `homer_preparse`. It prepare the genome for motif finding. Each entry has 3 main parameters: dbkey, maks (whether the small letters were masked), size.

Do you know if this format is changing frequently? In other DMs we are now starting to add a column for a "version" so that different versions of the tool could in theory use the same table and filter according to the "version" column.

> At the code levels, I have an interrogation:
> 
>     * data_manager table: I need to know which fasta file has been used to generate the preparse. I chose to store the full path in the table. I could have made a symlink and store only the basename. I don't know what is best.

I think the full path is correct.
 
> There are plenty of other tools from the homer suite to implement but a lot of them would require a new datatype as proposed here: https://github.com/bgruening/galaxytools/blob/master/tools/homer/datatypes_conf.xml

I'm not sure if I can find all the tools that have been requested over the years, but there is def. interest in a few more tools.

> Could you let me know if you are interested in supporting this tool suite?

I would say yes! :)

File: data_managers/data_manager_homer_preparsed/data_manager/homer_genome_preparse.xml [hunk 1]
```diff
@@ -0,0 +1,59 @@
+<tool id="homer_genome_preparse_data_manager" name="Homer genome preparse" tool_type="manage_data" version="@WRAPPER_VERSION@" profile="18.09">
+    <requirements>
+        <requirement type="package" version="@WRAPPER_VERSION@">homer</requirement>
+        <requirement type="package" version="3.7">python</requirement>
```
Comment (bgruening) [inline data_managers/data_manager_homer_preparsed/data_manager/homer_genome_preparse.xml:4]: 3.9?

File: data_managers/data_manager_homer_preparsed/data_manager/homer_genome_preparse.xml [hunk 2]
```diff
@@ -0,0 +1,59 @@
+<tool id="homer_genome_preparse_data_manager" name="Homer genome preparse" tool_type="manage_data" version="@WRAPPER_VERSION@" profile="18.09">
```
Comment (bgruening) [inline data_managers/data_manager_homer_preparsed/data_manager/homer_genome_preparse.xml:1]: profile=20.05?

File: tools/homer/homer_annotatePeaks.xml [hunk 3]
```diff
@@ -0,0 +1,133 @@
+<tool id="homer_annotatePeaks" name="annotatePeaks" version="@TOOL_VERSION@+galaxy@VERSION_SUFFIX@" python_template_version="3.5">
```
Comment (bgruening) [inline tools/homer/homer_annotatePeaks.xml:1]: remove python_template_version, please add profile=

File: tools/homer/homer_annotatePeaks.xml [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+  <macros>
+    <import>macros.xml</import>
+  </macros>
+  <expand macro="requirements"/>
+  <command detect_errors="exit_code"><![CDATA[
+				annotatePeaks.pl
```
Comment (bgruening) [inline tools/homer/homer_annotatePeaks.xml:7]: indentation looks off here

File: tools/homer/homer_annotatePeaks.xml [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+				## Peak file
+				$input
```
Comment (bgruening) [inline tools/homer/homer_annotatePeaks.xml:9]: single quote

File: tools/homer/homer_annotatePeaks.xml [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+						'${CpG.genome.all_fasta_source.fields.path}'
+					#else
+						'${CpG.genome.fasta}'
+					#end if
+					-CpG
+				#else
+					none
+				#end if
+				#if str($gtf_file) != "None":
+					-gtf $gtf_file
```
Comment (bgruening) [inline tools/homer/homer_annotatePeaks.xml:22]: single quote

File: tools/homer/homer_annotatePeaks.xml [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+				#end if
+				#if str($ann_file) != "None":
+					-ann $ann_file
+				#end if
+				> $output
```
Comment (bgruening) [inline tools/homer/homer_annotatePeaks.xml:27]: single qutoe

File: tools/homer/homer_annotatePeaks.xml [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+						'${CpG.genome.fasta}'
+					#end if
+					-CpG
+				#else
+					none
+				#end if
+				#if str($gtf_file) != "None":
+					-gtf $gtf_file
+				#end if
+				#if str($ann_file) != "None":
```
Comment (bgruening) [inline tools/homer/homer_annotatePeaks.xml:24]: `#if $ann_file` is probably enough

**Comment by lldelisle** [comment]
For me it seems that Homer is not maintained/updated anymore. The last update was (v4.11, 10-24-2019). I emailed the homer developer one month ago and did not get any news. I am afraid he does not have time to work on this since he opened his lab. But maybe we can add the version like in star just in case someone takes the project. Tell me what you think.

File: tools/homer/homer_gtf_to_annotation.xml [hunk 9]
```diff
@@ -0,0 +1,83 @@
+<tool id="homer_gtf_to_annotations" name="Converts GTF to Annotations file for Homer" version="@TOOL_VERSION@+galaxy0" python_template_version="3.5">
```
Comment (bgruening) [inline tools/homer/homer_gtf_to_annotation.xml:1]: same as above

File: tools/homer/macros.xml [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+      <when value="given"/>
+      <when value="fixed">
+        <param argument="-size" type="integer" min="0" value="200" label="fragment size in bp"/>
+        <expand macro="choose_center"/>
+      </when>
+    </conditional>
+  </xml>
+  <xml name="citation">
+    <citations>
+      <citation type="bibtex">
```
Comment (bgruening) [inline tools/homer/macros.xml:42]: can you use the DOI here?

**Review by bgruening — COMMENTED** [review]
@lldelisle please feel free to adjust/fix/replace the older tools.

**Comment by bgruening** [comment]
> For me it seems that Homer is not maintained/updated anymore. The last update was (v4.11, 10-24-2019). I emailed the homer developer one month ago and did not get any news. I am afraid he does not have time to work on this since he opened his lab. But maybe we can add the version like in star just in case someone takes the project. Tell me what you think.

I think a version does not hurt and should be future-proof.

File: data_managers/data_manager_homer_preparsed/data_manager/homer_genome_preparse.xml (hunk 2 above)
Comment (lldelisle) [reply data_managers/data_manager_homer_preparsed/data_manager/homer_genome_preparse.xml:1]: Is it mandatory? I would like that old galaxy can still install it. I don't know what changed at the tool level between 18.09 and 20.05.

**Comment by lldelisle** [comment]
Thanks for the review. I will include the proposed changes.

File: data_managers/data_manager_homer_preparsed/data_manager/homer_genome_preparse.xml (hunk 2 above)
Comment (bgruening) [reply data_managers/data_manager_homer_preparsed/data_manager/homer_genome_preparse.xml:1]: Not mandatory, but you will get a few defaults during the tool submissions differently (HOME, etc ...).

Comment (bgruening) [reply data_managers/data_manager_homer_preparsed/data_manager/homer_genome_preparse.xml:1]: But please note that one tool has 20.01 as profile already.

File: tools/homer/.shed.yml [hunk 11]
```diff
@@ -0,0 +1,16 @@
+categories:
+- Sequence Analysis
```
Comment (bgruening) [inline tools/homer/.shed.yml:2]: this file needs to be adopted and converted into a suite-shed.yml file.

File: data_managers/data_manager_homer_preparsed/data_manager/homer_genome_preparse.xml (hunk 2 above)
Comment (lldelisle) [reply data_managers/data_manager_homer_preparsed/data_manager/homer_genome_preparse.xml:1]: OK let's put 20.05

File: tools/homer/.shed.yml (hunk 11 above)
Comment (lldelisle) [reply tools/homer/.shed.yml:2]: I took example from bedtools which apprently is not a suite...

Comment (bgruening) [reply tools/homer/.shed.yml:2]: Yes, bedtools predates the suite stuff, unfortunately, and it's not easy to decompose now without breaking the tool linage.

Comment (lldelisle) [reply tools/homer/.shed.yml:2]: OK

File: data_managers/data_manager_homer_preparsed/data_manager/homer_genome_preparse.xml (hunk 1 above)
Comment (lldelisle) [reply data_managers/data_manager_homer_preparsed/data_manager/homer_genome_preparse.xml]: done

File: data_managers/data_manager_homer_preparsed/data_manager/homer_genome_preparse.xml (hunk 2 above)
Comment (lldelisle) [reply data_managers/data_manager_homer_preparsed/data_manager/homer_genome_preparse.xml]: done

File: tools/homer/homer_annotatePeaks.xml (hunk 4 above)
Comment (lldelisle) [reply tools/homer/homer_annotatePeaks.xml]: I don' know why there were tabulations... Thanks

File: tools/homer/homer_annotatePeaks.xml (hunk 5 above)
Comment (lldelisle) [reply tools/homer/homer_annotatePeaks.xml]: done

File: tools/homer/homer_annotatePeaks.xml (hunk 7 above)
Comment (lldelisle) [reply tools/homer/homer_annotatePeaks.xml]: done

File: tools/homer/homer_annotatePeaks.xml (hunk 8 above)
Comment (lldelisle) [reply tools/homer/homer_annotatePeaks.xml]: Yes indeed.

File: tools/homer/homer_gtf_to_annotation.xml (hunk 9 above)
Comment (lldelisle) [reply tools/homer/homer_gtf_to_annotation.xml]: done

File: tools/homer/macros.xml (hunk 10 above)
Comment (lldelisle) [reply tools/homer/macros.xml]: done

File: tools/homer/homer_annotatePeaks.xml (hunk 3 above)
Comment (lldelisle) [reply tools/homer/homer_annotatePeaks.xml]: done

File: tools/homer/homer_annotatePeaks.xml (hunk 6 above)
Comment (lldelisle) [reply tools/homer/homer_annotatePeaks.xml]: done

File: tools/homer/.shed.yml (hunk 11 above)
Comment (lldelisle) [reply tools/homer/.shed.yml:2]: Done

**Comment by lldelisle** [comment]
For the `homer_findMotifsGenome`, the 2 mains outputs are the html I put but then there are additional files like the `*.motif` files which could be use for further analysis so I don't know if they should be displayed as output like a collection. For the moment the user can download them as they are part of the files associated to the html so when they download the html they get all the associated files.

File: .gitignore [hunk 12]
```diff
@@ -63,9 +63,9 @@ target/
...
 
+# Tool test from planemo
 tool_test_output.html
 tool_test_output.json
 
-.DS_Store
-
-
+# Tool test from galaxy-language-server
+*_test_report.html
```
Comment (bernt-matthias) [inline .gitignore:71]: Did you intended to change this file?

File: data_managers/data_manager_homer_preparsed/test-data/all_fasta.loc [hunk 13]
```diff
@@ -0,0 +1,19 @@
+#This file lists the locations and dbkeys of all the fasta files
+#under the "genome" directory (a directory that contains a directory
```
Comment (bernt-matthias) [inline data_managers/data_manager_homer_preparsed/test-data/all_fasta.loc:2]: Can't remember: is this file needed, or just the sample and test file?

File: outputs/CTCF_peaks_second.txt [hunk 14]
```diff
@@ -0,0 +1,67 @@
+PeakID (cmd=annotatePeaks.pl /tmp/tmpmsg4l31h/files/1/b/4/dataset_1b4fbb9d-2adf-4d0b-8495-abe5623141ad.dat none -ann /tmp/tmpmsg4l31h/files/a/f/d/dataset_afd05a45-6170-414d-942f-e4a8610f65b4.dat)	Chr	Start	End	Strand	Peak Score	Focus Ratio/Region Size	Annotation	Detailed Annotation	Distance to TSS	Nearest PromoterID	Entrez ID	Nearest Unigene	Nearest Refseq	Nearest Ensembl	Gene Name	Gene Alias	Gene Description	Gene Type
```
Comment (bernt-matthias) [inline outputs/CTCF_peaks_second.txt:1]: Is this directory and the files in it needed?

File: tools/homer/homer_annotatePeaks.xml [hunk 15]
```diff
@@ -0,0 +1,133 @@
...
+            </param>
+          </when>
+          <when value="history">
+            <param name="fasta" type="data" format="fasta" label="Select reference genome"/>
+          </when>
+        </conditional>
+      </when>
+      <when value="False"/>
+    </conditional>
+    <param name="gtf_file" type="data" format="gtf,gff,gff3" optional="True" label="GTF file" help="Must be of datatype &quot;gtf&quot; or &quot;gff&quot; or &quot;gff3&quot;" argument="-gtf"/>
```
Comment (bernt-matthias) [inline tools/homer/homer_annotatePeaks.xml:58]: Better use lower case `true` and `false`.

File: tools/homer/homer_findMotifsGenome.xml [hunk 16]
```diff
@@ -0,0 +1,411 @@
...
+      </conditional>
+      <output name="html_file" file="motif_test1/knownResults.html" ftype="html" lines_diff="2" />
+      <output name="html_homer_file" >
+        <assert_contents>
+            <has_text text="fake_phix_peaks_bed_motif/ - Homer de novo Motif Results"/>
+            <has_text text="Total target sequences = 1"/>
+            <has_text text="Jaspar"/>
+        </assert_contents>
+      </output>
+  </test>
```
Comment (bernt-matthias) [inline tools/homer/homer_findMotifsGenome.xml:245]: Inconsistent indentation.

Btw IUC prefers 4 spaces.

File: .gitignore (hunk 12 above)
Comment (lldelisle) [reply .gitignore:71]: Yes I think galaxy-language-server was written to help wrapper writting and that's why I thought it would be a good idea to modify the .gitignore.

File: outputs/CTCF_peaks_second.txt (hunk 14 above)
Comment (lldelisle) [reply outputs/CTCF_peaks_second.txt]: Oups good catch

File: data_managers/data_manager_homer_preparsed/test-data/all_fasta.loc (hunk 13 above)
Comment (lldelisle) [reply data_managers/data_manager_homer_preparsed/test-data/all_fasta.loc:2]: The test fail if I remove this file.

File: tools/homer/homer_findMotifsGenome.xml (hunk 16 above)
Comment (lldelisle) [reply tools/homer/homer_findMotifsGenome.xml]: Done

File: tools/homer/homer_annotatePeaks.xml (hunk 15 above)
Comment (lldelisle) [reply tools/homer/homer_annotatePeaks.xml]: Done

**Comment by lldelisle** [comment]
Thanks @bernt-matthias for the review.

**Review by bernt-matthias — APPROVED** [review]

**Review by bgruening — APPROVED** [review]
Thanks!

**Comment by bgruening** [comment]
@lldelisle I have ask our users and they are missing the HOMER motif tools http://homer.ucsd.edu/homer/motif/index.html

Is that something that is also interesting for you?

**Comment by lldelisle** [comment]
@bgruening I wrapped `findMotifsGenome.pl` which is the motif tool I use (motif finding from peaks coordinates and fasta). They want the `findMotifs.pl` (motif finding from fasta)?

**Comment by mtekman** [comment]
very excited to see this tool on Galaxy -- it is exactly what I was looking for!
